### PR TITLE
Integrate NGINX App Protect DoS

### PR DESCRIPTION
**Pull request by pleshakov**
This PR brings support for NGINX App Protect DoS

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -119,3 +127,8 @@ deps: ## Add missing and remove unused modules, verify deps and dowload them to
 .PHONY: clean-cache
 clean-cache: ## Clean go cache
 	@go clean -modcache
+
+.PHONY: app-protect-dos-arb
```
[the remark this answers is not in the document]
Comment (pasmant) [reply Makefile:133]: removed

File: build/Dockerfile [hunk 2]
```diff
@@ -218,7 +250,7 @@ ARG GIT_COMMIT
 ARG DATE
 ARG TARGETPLATFORM
 
-RUN mkdir -p /var/lib/nginx /etc/nginx/secrets /etc/nginx/stream-conf.d \
+RUN mkdir -p /etc/nginx/dos/policies /etc/nginx/dos/logconfs /var/lib/nginx /etc/nginx/secrets /etc/nginx/stream-conf.d \
```
[the remark this answers is not in the document]
Comment (pasmant) [reply build/Dockerfile:253]: Done

File: cmd/nginx-ingress/main.go [hunk 3]
```diff
@@ -548,6 +555,13 @@ func main() {
 		nginxManager.CreateTLSPassthroughHostsConfig(emptyFile)
 	}
 
+	var aPPDosAgentDone chan error
+
+	if *appProtectDos {
+		aPPDosAgentDone = make(chan error, 1)
+		nginxManager.AppProtectDosAgentStart(aPPDosAgentDone, *nginxDebug || cfgParams.AppProtectDosDebug, cfgParams.AppProtectDosMaxDaemon, cfgParams.AppProtectDosMaxWorkers, cfgParams.AppProtectDosMemory) // Add Debug bool option via config file
```
[the remark this answers is not in the document]
Comment (pasmant) [reply cmd/nginx-ingress/main.go:562]: Will remove the comment

File: internal/configs/version1/nginx-plus.ingress.tmpl [hunk 4]
```diff
@@ -65,15 +65,30 @@ server {
 	set $resource_namespace "{{$.Ingress.Namespace}}";
 
 	{{- if $server.AppProtectEnable}}
-	app_protect_enable {{$server.AppProtectEnable}};
+	app_protect_enable {{$server.AppProtectEnable}};{{- end}}
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/version1/nginx-plus.ingress.tmpl:68]: Will fix it, no need to add this at the first place

File: build/Dockerfile [hunk 5]
```diff
@@ -80,6 +80,38 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
...
+RUN mkdir -p /shared/cores /var/log/adm /etc/app_protect_dos /var/run/adm/ && chmod 777 /shared/cores /var/log/adm /etc/app_protect_dos /var/run/adm/
+
+RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode=0644 \
+    --mount=type=secret,id=nginx-repo.key,dst=/etc/ssl/nginx/nginx-repo.key,mode=0644 \
+    set -x \
+    && apt-get update \
+    && apt-get -y install lsb-release ca-certificates \
+    && printf "deb https://pkgs.nginx.com/app-protect-dos/debian `lsb_release -cs` nginx-plus\n" | tee /etc/apt/sources.list.d/nginx-app-protect-dos.list \
+    && apt-get update \
+    && apt-get -y install app-protect-dos
```
[the remark this answers is not in the document]
Comment (pasmant) [reply build/Dockerfile:97]: done

File: internal/configs/version1/nginx-plus.ingress.tmpl [hunk 6]
```diff
@@ -65,15 +65,30 @@ server {
...
+	{{if $server.AppProtectDosLogConf}}app_protect_dos_security_log {{$server.AppProtectDosLogConf}};{{end}}
+	{{- end}}
+	{{- if $server.AppProtectDosAccessLogDst}}
+	set $loggable '0';
+	# app-protect-dos module will set it to '1'  if a request doesn't pass the rate limit
+	access_log syslog:server={{ .AppProtectDosAccessLogDst }} log_dos if=$loggable;
+	{{- end}}
+	{{if $server.AppProtectDosMonitor}}app_protect_dos_monitor "{{$server.AppProtectDosMonitor}}";{{end}}
+	{{if $server.AppProtectDosName}}app_protect_dos_name "{{$server.AppProtectDosName}}";{{end}}
+    {{- end}}
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/version1/nginx-plus.ingress.tmpl:90]: done

File: internal/configs/annotations.go [hunk 7]
```diff
@@ -31,21 +40,29 @@ var masterBlacklist = map[string]bool{
...
+	"nginx.org/server-snippets":                                     true,
+	"appprotect.f5.com/app_protect_enable":                          true,
+	"appprotect.f5.com/app_protect_policy":                          true,
+	"appprotect.f5.com/app_protect_security_log_enable":             true,
+	"appprotect.f5.com/app_protect_security_log":                    true,
+	"appprotectdos.f5.com/app_protect_dos_enable":                   true,
+	"appprotectdos.f5.com/app_protect_dos_policy":                   true,
+	"appprotectdos.f5.com/app_protect_dos_security_log_enable":      true,
+	"appprotectdos.f5.com/app_protect_dos_security_log":             true,
+	"appprotectdos.f5.com/app_protect_dos_security_log_destination": true,
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/annotations.go:62]: Here is blacklist with underscore _ (not like the const at the top)

File: internal/configs/configurator.go [hunk 8]
```diff
@@ -1484,6 +1443,32 @@ func (cnf *Configurator) RefreshAppProtectUserSigs(
...
+// DeleteAppProtectDosPolicy updates Ingresses and VirtualServers that use AP Dos Policy after that policy is deleted
+func (cnf *Configurator) DeleteAppProtectDosPolicy(resource *unstructured.Unstructured, polNamespaceName string, ingExes []*IngressEx, mergeableIngresses []*MergeableIngresses, vsExes []*VirtualServerEx) (Warnings, error) {
+	if len(ingExes)+len(mergeableIngresses)+len(vsExes) > 0 {
+		cnf.nginxManager.DeleteAppProtectResourceFile(appProtectResourceFromNsName(appProtectDosPolicyFolder, polNamespaceName))
+	}
+
+	return  cnf.AddOrUpdateAppProtectResource(resource, ingExes, mergeableIngresses, vsExes, true)
+}
+
+// DeleteAppProtectLogConf updates Ingresses and VirtualServers that use AP Log Configuration after that policy is deleted
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/configurator.go:1463]: done

File: internal/configs/annotations.go [hunk 9]
```diff
@@ -372,6 +389,43 @@ func parseAnnotations(ingEx *IngressEx, baseCfgParams *ConfigParams, isPlus bool
 			}
 		}
 
+	}
+	if hasAppProtectDos {
+		if appProtectDosEnable, exists, err := GetMapKeyAsBool(ingEx.Ingress.Annotations, "appprotectdos.f5.com/app-protect-dos-enable", ingEx.Ingress); exists {
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/annotations.go:394]: the annotations here not appear in the top. the top const here used in other file

File: internal/configs/version2/nginx-plus.virtualserver.tmpl [hunk 10]
```diff
@@ -331,6 +357,32 @@ server {
...
+        app_protect_dos_name "{{ .Name }}";
+            {{- end }}
+
+            {{- if .ApDosPolicy }}
+        app_protect_dos_policy_file {{ .ApDosPolicy }};
+            {{- end }}
+
+            {{ if .ApDosSecurityLogEnable }}
+        app_protect_dos_security_log_enable on;
+        app_protect_dos_security_log {{ .ApDosLogConf }};
```
[the remark this answers is not in the document]
Comment (aknot242) [reply internal/configs/version2/nginx-plus.virtualserver.tmpl:373]: @pleshakov the combination of both logs should be used if the customer wants to use the [NAP DoS dashboard](https://github.com/f5devcentral/nap-dos-elk-dashboards#nginx-app-protect-dos-configuration). Some of the logs are provided by nginx access logs over TCP (using custom variables), and some of the logs are from DoS itself (over UDP). There may use cases other than the dashboard.

File: internal/configs/configurator.go [hunk 11]
```diff
@@ -1301,6 +1317,24 @@ func (cnf *Configurator) updateApResourcesForVs(vsEx *VirtualServerEx) map[strin
 		}
 	}
 
+	if vsEx.ApDosPolRefs != nil {
```
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/configurator.go:1320]: https://github.com/nginxinc/kubernetes-ingress/pull/2032

File: internal/k8s/appprotectdos/app_protect_dos_configuration.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+	AddOrUpdatePolicy(policyObj *unstructured.Unstructured) (changes []Change, problems []Problem)
+	AddOrUpdateLogConf(logConfObj *unstructured.Unstructured) (changes []Change, problems []Problem)
+	GetAppResource(kind, key string) (*unstructured.Unstructured, error)
+	DeletePolicy(key string) (changes []Change, problems []Problem)
+	DeleteLogConf(key string) (changes []Change, problems []Problem)
+}
+
+// ConfigurationImpl holds representations of App Protect Dos cluster resources
+type ConfigurationImpl struct {
+	DosPolicies map[string]*DosPolicyEx
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_configuration.go:85]: Make them private

File: internal/k8s/appprotectdos/app_protect_dos_resources.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+	if net.ParseIP(dstchunks[0]) == nil {
+		return fmt.Errorf("Error parsing host: %v is not a valid ip address", dstchunks[0])
+	}
+
+	return nil
+}
+
+// ValidateAppProtectDosName validates name of App Protect Dos
+func ValidateAppProtectDosName(name string) error {
+	if (len(name) > MaxNameLength) {
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go:65]: NGINX module make only length validation on the name
The directive not support variables
For the example above the vs_name will be hello$host.
Took from admd:
VS Stats:
vs_name:                                 hello$host
vs_hash:                                 3751576534
vs_id:                                   1
vs_uri:                                  webapp.example.com

File: internal/k8s/controller.go [hunk 14]
```diff
@@ -1239,6 +1278,88 @@ func (lbc *LoadBalancerController) processAppProtectProblems(problems []appprote
...
+				name := impl.Obj.GetName()
+				resources := lbc.configuration.FindResourcesForAppProtectDosLogConfAnnotation(namespace, name)
+
+				for _, wafPol := range getBadosPoliciesForAppProtectLogConf(lbc.getAllPolicies(), namespace+"/"+name) {
+					resources = append(resources, lbc.configuration.FindResourcesForPolicy(wafPol.Namespace, wafPol.Name)...)
+				}
+
+				resourceExes := lbc.createExtendedResources(resources)
+
+				warnings, deleteErr := lbc.configurator.DeleteAppProtectLogConf(impl.Obj, namespace+"/"+name, resourceExes.IngressExes, resourceExes.MergeableIngresses, resourceExes.VirtualServerExes)
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/controller.go:1346]: Fix, call the DeleteAppProtectDosLogConf

File: internal/k8s/appprotectdos/app_protect_dos_resources.go [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+		return fmt.Errorf("App Protect Dos Name max length is %v", MaxNameLength)
+	}
+
+	return nil
+}
+
+// ValidateAppProtectDosMonitor validates monitor value of App Protect Dos
+func ValidateAppProtectDosMonitor(monitor string) error {
+	_, err := url.Parse(monitor)
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go:74]: NGINX module not make any validation
The directive not support variables
For the example above the vs_name will be hello$host.
Took from admd:
VS Stats:
```
vs_name:                                 hello$host
vs_hash:                                 3751576534
vs_id:                                   1
vs_uri:                                  hello$host
```

File: deployments/helm-chart-dos-arbitrator/values.yaml [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  resources: 
+    limits:
+      cpu: 500m
+      memory: 128Mi
+
+  image:
+    ## The image repository of the Arbitrator.
+    repository: docker-registry.nginx.com/nap-dos/app_protect_dos_arb
+
+    ## The tag of the Ingress controller image.
```
[the remark this answers is not in the document]
Comment (pasmant) [reply deployments/helm-chart-dos-arbitrator/values.yaml:12]: done

File: deployments/helm-chart-dos-arbitrator/values.yaml [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    tag: "latest"
+
+    ## The pull policy for the Ingress controller image.
```
[the remark this answers is not in the document]
Comment (pasmant) [reply deployments/helm-chart-dos-arbitrator/values.yaml:15]: Done

File: internal/k8s/validation_test.go [hunk 18]
```diff
@@ -1393,12 +1500,280 @@ func TestValidateNginxIngressAnnotations(t *testing.T) {
...
+			annotations: map[string]string{
+				"appprotectdos.f5.com/app-protect-dos-security-log-destination": "stderr",
+			},
+			specServices:          map[string]bool{},
+			isPlus:                true,
+			appProtectEnabled:     false,
+			appProtectDosEnabled:  true,
+			internalRoutesEnabled: false,
+			expectedErrors: nil,
+			msg: "valid app-protect-dos-security-log-destination annotation",
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/validation_test.go:1693]: done

File: internal/k8s/validation.go [hunk 19]
```diff
@@ -353,6 +410,29 @@ func validateRelatedAnnotation(name string, validator validatorFunc) annotationV
...
+	}
+}
+
+func validateQualifiedName(context *annotationValidationContext) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	err := validation.IsQualifiedName(context.value)
+
+	if err != nil {
+		return append(allErrs, field.Invalid(context.fieldPath, context.value, fmt.Sprintf("annotation value: %v is not qualified name", context.value)))
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/validation.go:430]: done also for unit test

File: deployments/helm-chart-dos-arbitrator/templates/controller-appprotect-dos-arb.yaml [hunk 20]
```diff
@@ -0,0 +1,40 @@
+apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (pasmant) [reply deployments/helm-chart-dos-arbitrator/templates/controller-appprotect-dos-arb.yaml:1]: done, make yaml file to deployment and one for the svc

File: internal/configs/version1/nginx-plus.tmpl [hunk 21]
```diff
@@ -51,6 +54,10 @@ http {
                       '"$http_user_agent" "$http_x_forwarded_for"';
     {{- end}}
 
+    {{- if and .AppProtectDosLoadModule (.AppProtectDosLogFormat)}}
+    log_format log_dos '{{.AppProtectDosLogFormat}}';
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/version1/nginx-plus.tmpl:58]: Add default value for log_dos

File: deployments/helm-chart-dos-arbitrator/Chart.yaml [hunk 22]
```diff
@@ -0,0 +1,17 @@
+name: nginx-appprotect-dos-arbitrator
+version: 0.0.1
+appVersion: 1.12.1
```
[the remark this answers is not in the document]
Comment (pasmant) [reply deployments/helm-chart-dos-arbitrator/Chart.yaml:3]: done

File: pkg/apis/configuration/validation/policy.go [hunk 23]
```diff
@@ -277,13 +295,59 @@ func validateLogConf(logConf, logDest string, fieldPath *field.Path) field.Error
...
 
-	err := appprotect.ValidateAppProtectLogDestination(logDest)
+	err := appprotect_common.ValidateAppProtectLogDestination(logDest)
 	if err != nil {
 		allErrs = append(allErrs, field.Invalid(fieldPath.Child("logDest"), logDest, err.Error()))
 	}
 	return allErrs
 }
 
+func validateBados(bados *v1.Bados, fieldPath *field.Path) field.ErrorList {
```
[the remark this answers is not in the document]
Comment (pasmant) [reply pkg/apis/configuration/validation/policy.go:305]: done

File: pkg/apis/configuration/validation/policy.go [hunk 24]
```diff
@@ -277,13 +295,59 @@ func validateLogConf(logConf, logDest string, fieldPath *field.Path) field.Error
...
+	allErrs := field.ErrorList{}
+
+	if bados.ApDosPolicy != "" {
+		for _, msg := range validation.IsQualifiedName(bados.ApDosPolicy) {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("apDosPolicy"), bados.ApDosPolicy, msg))
+		}
+	}
+
+    if bados.DosAccessLogDest != "" {
```
[the remark this answers is not in the document]
Comment (pasmant) [reply pkg/apis/configuration/validation/policy.go:314]: done

File: pkg/apis/configuration/validation/policy.go [hunk 25]
```diff
@@ -100,10 +101,27 @@ func validatePolicySpec(spec *v1.PolicySpec, fieldPath *field.Path, isPlus, enab
 		fieldCount++
 	}
 
+	if spec.Bados != nil {
```
[the remark this answers is not in the document]
Comment (pasmant) [reply pkg/apis/configuration/validation/policy.go:104]: done, also added for TestValidatePolicy

File: internal/configs/configurator.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/configurator.go]: merged and done

File: internal/configs/configurator_test.go [hunk 26]
```diff
@@ -1283,7 +1283,204 @@ func TestUpdateApResourcesForVs(t *testing.T) {
 	for _, test := range tests {
 		result := conf.updateApResourcesForVs(test.vsEx)
 		if !reflect.DeepEqual(result, test.expected) {
-			t.Errorf("updateApResourcesForVs() returned \n%v but exexpected\n%v for the case of %s", result, test.expected, test.msg)
+			t.Errorf("updateApResourcesForVs() returned \n%v but expected\n%v for the case of %s", result, test.expected, test.msg)
+		}
+	}
+}
+
+func TestUpdateApDosResources(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/configurator_test.go:1306]: this is no longer relevant after the refactoring done in https://github.com/nginxinc/kubernetes-ingress/pull/2032

File: internal/configs/annotations.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/annotations.go:62]: oh, this is a bug. We will fix the NAP annotations separately. Could you fix the DOS-related annotations?

File: internal/k8s/appprotectdos/app_protect_dos_resources.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go:65]: in this case, on top of the max length validation, we need to insure that the value inside cannot escape "..." string, which can lead to a failed NGINX reload, which would be similar to https://github.com/nginxinc/kubernetes-ingress/blob/4688555a3e0337a9fbac4053d159d49f232d0940/pkg/apis/configuration/validation/virtualserver.go#L796-L799

File: internal/configs/configurator.go [hunk 27]
```diff
@@ -1351,121 +1390,41 @@ func (cnf *Configurator) AddOrUpdateAppProtectResource(resource *unstructured.Un
...
-		return allWarnings, fmt.Errorf("Error when reloading NGINX when removing App Protect Policy: %w", err)
-	}
-
-	return allWarnings, nil
+	return cnf.AddOrUpdateAppProtectResource(resource, ingExes, mergeableIngresses, vsExes, true)
 }
 
 // DeleteAppProtectLogConf updates Ingresses and VirtualServers that use AP Log Configuration after that policy is deleted
-func (cnf *Configurator) DeleteAppProtectLogConf(logConfNamespaceName string, ingExes []*IngressEx, mergeableIngresses []*MergeableIngresses, vsExes []*VirtualServerEx) (Warnings, error) {
+func (cnf *Configurator) DeleteAppProtectLogConf(resource *unstructured.Unstructured, logConfNamespaceName string, ingExes []*IngressEx, mergeableIngresses []*MergeableIngresses, vsExes []*VirtualServerEx) (Warnings, error) {
```
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/configurator.go:1421]: addressed in https://github.com/nginxinc/kubernetes-ingress/pull/2036

File: internal/configs/configurator.go [hunk 28]
```diff
@@ -1323,8 +1357,13 @@ func generateApResourceFileContent(apResource *unstructured.Unstructured) []byte
...
 }
 
+func appProtectResourceFromNsName(folderName string, polNamespaceName string) string {
+	fName := strings.Replace(polNamespaceName, "/", "_", 1)
+	return folderName + fName
+}
+
 // AddOrUpdateAppProtectResource updates Ingresses and VirtualServers that use App Protect Resources
-func (cnf *Configurator) AddOrUpdateAppProtectResource(resource *unstructured.Unstructured, ingExes []*IngressEx, mergeableIngresses []*MergeableIngresses, vsExes []*VirtualServerEx) (Warnings, error) {
+func (cnf *Configurator) AddOrUpdateAppProtectResource(resource *unstructured.Unstructured, ingExes []*IngressEx, mergeableIngresses []*MergeableIngresses, vsExes []*VirtualServerEx, reload bool) (Warnings, error) {
```
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/configurator.go:1375]: addressed in https://github.com/nginxinc/kubernetes-ingress/pull/2036

File: internal/configs/annotations.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/annotations.go:394]: ok. please disregard this request

File: internal/configs/annotations.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pleshakov) [reply internal/configs/annotations.go:62]: addressed in https://github.com/nginxinc/kubernetes-ingress/pull/2037

File: internal/k8s/appprotectdos/app_protect_dos_resources.go [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+var appProtectDosPolicyRequiredFields = [][]string{
+	{"spec"},
+}
+
+var appProtectDosLogConfRequiredFields = [][]string{
+	{"spec", "content"},
+	{"spec", "filter"},
+}
+
+const MaxNameLength = 64
```
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go:25]: Add doc inside the policy-resource.md and in advanced-configuration-with-annotations.md

File: internal/k8s/appprotectdos/app_protect_dos_resources.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go]: Add new validation

File: internal/k8s/appprotectdos/app_protect_dos_resources.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go]: Add new validation

File: internal/configs/version1/nginx-plus.tmpl (hunk 21 above)
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/configs/version1/nginx-plus.tmpl:58]: Will fix in PR #1999

File: internal/k8s/appprotectdos/app_protect_dos_resources.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pasmant) [reply internal/k8s/appprotectdos/app_protect_dos_resources.go]: https://github.com/nginxinc/kubernetes-ingress/pull/2076

File: pkg/apis/configuration/validation/dos.go [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+	if err := validateEscapedString(name, "protected-object-one"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateAppProtectDosMonitor validates monitor value of App Protect Dos
+func ValidateAppProtectDosMonitor(monitor string) error {
+	_, err := url.Parse(monitor)
```
Comment (aknot242) [inline pkg/apis/configuration/validation/dos.go:81]: In NAP DoS 2.0, the `app_protect_dos_monitor` directive has changed. It now supports `uri`, `protocol` and `timeout` fields. In gRPC streaming use cases, the `uri` and `protocol` fields are required to be set. Therefore, validation by simply parsing as a URL will no longer be sufficient. http://localhost:1313/nginx-app-protect-dos/directives-and-policy/learn-about-directives-and-policy/#for-nginx-app-protect-dos-v2

**Review by aknot242 — CHANGES_REQUESTED** [review]

**Comment by soneillf5** [comment]
Closing this pull request and re-opening it. We've addressed the comments made and would like a less noisy PR to do a fresh review.

**Pull request closed without merging**